### PR TITLE
feat: BSP skipping unconfirmed files

### PR DESCRIPTION
**Pull request by snowmead**
BSPs can batch confirm multiple files after volunteering for them. But since BSPs are essentially racing to confirm for the same files if they pass the volunteering threshold, it is possible for a BSP to send a confirmation transaction after another BSP's transaction (resulting in the storage request either reaching the `ReplicationTarget` or the storage request being fulfilled (storage request is no longer in storage / non-existant) within the same block which would fail the transaction. 

Instead of failing the transaction, we simply skip it and emit in the `BspConfirmedStoring` event any skipped file keys. Now in the BSP task handling result of this transaction, it reacts and filters out any skipped file keys locally and apply only the successfully confirmed ones to the local forest.

### CODE REVIEW

**Review by ffarall — APPROVED** [review]
Great job! Just a few comments that I trust you'll get out of the way quickly. Therefore I approve.

File: pallets/file-system/src/utils.rs [hunk 1]
```diff
@@ -1148,16 +1148,22 @@ where
...
+        // Check if the BSP is eligible to volunteer for the storage request.
+        let current_tick_number =
+            <T::ProofDealer as shp_traits::ProofsDealerInterface>::get_current_tick();
 
-        // Check that the BSP's threshold is under the threshold required to volunteer for the storage request.
-        ensure!(bsp_threshold <= to_succeed, Error::<T>::AboveThreshold);
+        ensure!(
+            current_tick_number >= earliest_volunteer_tick,
+            Error::<T>::BspNotEligibleToVolunteer
+        );
```
[the remark this answers is not in the document]
Comment (snowmead) [reply pallets/file-system/src/utils.rs:1171]: to keep it consistent with the rest of the codebase, and it's simpler to read

File: pallets/file-system/src/utils.rs [hunk 2]
```diff
@@ -1407,45 +1413,78 @@ where
...
-        // Compute new root after inserting new file keys in forest partial trie.
-        let new_root = <T::ProofDealer as shp_traits::ProofsDealerInterface>::apply_delta(
-            &bsp_id,
+            // Update root of BSP.
+            <T::Providers as shp_traits::MutateProvidersInterface>::update_root(bsp_id, new_root)?;
+
+            new_root
+        } else {
+            return Err(Error::<T>::NoFileKeysToConfirm.into());
+        };
```
[the remark this answers is not in the document]
Comment (snowmead) [reply pallets/file-system/src/utils.rs:1467]: good suggestion

File: pallets/proofs-dealer/src/utils.rs [hunk 3]
```diff
@@ -1005,7 +1005,14 @@ impl<T: pallet::Config> ProofsDealerInterface for Pallet<T> {
             <T::ForestVerifier as TrieProofDeltaApplier<T::MerkleTrieHashing>>::apply_delta(
                 &root, mutations, proof,
             )
-            .map_err(|_| Error::<T>::FailedToApplyDelta)?
+            .map_err(|e| {
+                log::error!(
+                    "Failed to apply delta for Provider {:?} with error: {:?}",
+                    provider_id,
+                    e
+                );
```
[the remark this answers is not in the document]
Comment (snowmead) [reply pallets/proofs-dealer/src/utils.rs:1013]: the only reason I want this here specifcally wes to debug if something really goes wrong. should we keep it?

[the remark this answers is not in the document]
Comment (ffarall) [reply pallets/proofs-dealer/src/utils.rs:1013]: Summary: we want to keep the wasm blob small.

[the remark this answers is not in the document]
Comment (snowmead) [reply pallets/proofs-dealer/src/utils.rs:1013]: note: we should add an `Error` associated type which is returned by the apply delta and is pushed to the `FailedToApplyDelta` error which would be a tupple.

**Pull request merged**